### ISSUE TITLE
Support python array API

### ISSUE THREAD

**Issue by emcastillo** [pytorch/pytorch#54581]
## 🚀 Feature

Hi, we would like to discuss the PyTorch adoption of the recently proposed array-api: https://data-apis.org/array-api/latest/.
We are aware of the proposal in #54138 to adopt `__array_interface__` in the PyTorch tensor but its main purpose seems to be eliminating the current PyTorch and numpy dependency.

## Motivation

The array API is similar to the `__array_interface__` but extended beyond the realms of NumPy and providing support for accelerators. So it will ease the portability and reduce the burden of having to manage different interfaces such as DLPack or array interface depending on the device the array resides in.

We are wondering if support for this API would be welcomed in PyTorch and we would like to actively develop it since we are also planning to implement this in CuPy and both projects could benefit from this.

## Discussion points

One of the most important points to discuss is how would be adding support for this. Adding a submodule to ensure compatibility? in the same way that tensorflow added numpy compatibility `tensorflow.experimental.numpy`.

Or do we directly replace the current functions to be compatible with this API, but we break backward compatibility.

cc @mruberry @rgommers @heitorschueroff @albanD @kmaehashi @leofang @ezyang @apaszke

**Comment by rgommers** [comment]
_Full disclosure, I'm one of the authors of https://data-apis.org/array-api/latest/, together with @apaszke (PyTorch, JAX) and @leofang (CuPy)._

Thanks for this feature request @emcastillo! And great to see it coming from the CuPy team. I agree there's a lot of synergy and value here, and would like to collaborate on this.

Some thoughts:
- @mruberry, @ezyang and I have discussed the array API standardization effort several times, but don't have a concrete implementation plan yet.
- `torch.linalg` is already being designed with this standard in mind, and should be at or close to 100% compatibility.
- `__array_interface__` is only similar to one particular feature in the array API standard: `__dlpack__` + `__dlpack_device__`. I agree DLPack is much more interesting than `__array_interface__`, because it has support for accelerators. PyTorch already has support for DLPack in `torch.utils.dlpack`, but the API isn't great and it's too hidden. Getting all tensor/array libraries to implement the new DLPack interface so we finally have a single protocol that works across all libraries (including stream support) is definitely valuable, and probably the first bit to tackle. I'd like to open a separate feature request for that, since it is easy to do and makes sense as a standalone feature.
- The full array API standard is a large feature request, and it will need an implementation plan with multiple steps.
- There's a portable test suite for the standard, which should work with PyTorch out of the box: https://github.com/data-apis/array-api-tests. It'd be interesting to try, I expect the vast majority of the ~120 functions to already be compatible.

> One of the most important points to discuss is how would be adding support for this. Adding a submodule to ensure compatibility? in the same way that tensorflow added numpy compatibility `tensorflow.experimental.numpy`.
> 
> Or do we directly replace the current functions to be compatible with this API, but we break backward compatibility.

This is a great question. Let me also link to the [NumPy implementation plan, NEP 47](https://numpy.org/neps/nep-0047-array-api-standard.html). In NumPy we decided on a separate module, but the main reason was that we have to make more backwards-incompatible changes in NumPy than will be necessary in PyTorch and CuPy. Things like type casting behavior, returning 0-D arrays instead of scalars, etc. - PyTorch already gets this right.

I think it would be preferable to have compatibility in `torch.Tensor` and the `torch` namespace itself (+ `torch.linalg`, no other submodules will be involved). This should be possible. The main issue will probably be some signatures that have a slight divergence between the API standard and NumPy - so PyTorch cannot be compatible with both. It'd be useful to have a detailed comparison of those differences to make this decision.

**Comment by leofang** [comment]
ref: https://github.com/cupy/cupy/issues/4789 (CuPy's counterpart issue)

**Comment by emcastillo** [comment]
> It'd be useful to have a detailed comparison of those differences to make this decision.

we have started this on our side @asi1024 is working on summarizing the differences between pytorch and ndarray API :)

**Comment by ezyang** [comment]
In general we're happy to take patches that improve compatibility; we don't have anyone fulltiming on it from our side. Handling incompatibility is definitely a problem and something we'll have to hash out.

**Comment by asi1024** [comment]
We've investigated each API in the Data API specification and summarized modifications needed to run code written using the API on PyTorch.
Here is the full list.
https://docs.google.com/spreadsheets/d/13amsIeHhe38apCeZrOmsbsHT-aysmYur9-cugw5jVg0/edit#gid=0
In the list, we have classified each API into 6 categories:
* No Changes: The current PyTorch implementation already conforms to the Data API.
* New API: Need to add a new function or constant.
* Enhancement: Need enhancement to the existing function (add new argument, etc.).
* Enhancement with BC: Need to change signature of the existing function. (*breaking change* if we implement the API directly under `torch.*` namespace)
* Behavior: Need to change behavior of the existing function. (*breaking change* if we implement the API directly under `torch.*` namespace)
* N/A: The detailed specification is not determined yet in the Data API.

To implement the Data API directly under `torch.*` namespace, we need to discuss how to deal with the breaking changes. Major topics are:
- Arguments: Some API cannot be compatible with the Data API specification without breaking changes. e.g., `torch.linspace(start, end, steps, ...)` v.s. Data API `linspace(start, stop, num, ...)`.
- Argument names: PyTorch APIs conventionally use `dim`/`dims` as an argument name (e.g., `torch.stack`), whereas the Data API requires `axis` keyword argument.
- Return value: For example, `torch.nonzero` returns a tensor when `as_tuple=False` (default) whereas the Data API requires to always return as tuple.
- Behavior:
  - `torch.cross` looks for the first dimension found with the size 3, whereas the Data API uses -1 as default axis.
  - `floordiv` operator needs be changed to take floor. (The [documentation](https://pytorch.org/docs/stable/generated/torch.floor_divide.html) says `This behavior will be deprecated in a future PyTorch release` so it may be a good time to change the behavior)

**Comment by leofang** [comment]
btw there is also an API comparison tool, with data collected for NumPy, CuPy, PyTorch, etc:
https://github.com/data-apis/array-api-comparison
example comparison:
https://github.com/data-apis/array-api-comparison/tree/main/signatures/arithmetic

**Comment by ezyang** [comment]
Thank you for the analysis

> Arguments: Some API cannot be compatible with the Data API specification without breaking changes. e.g., torch.linspace(start, end, steps, ...) v.s. Data API linspace(start, stop, num, ...)

Is this actually an incompatibility?  I think steps and num have the same semantics, it's just kwarg remapping

> Argument names: PyTorch APIs conventionally use dim/dims as an argument name (e.g., torch.stack), whereas the Data API requires axis keyword argument.

Long standing issue, see https://github.com/pytorch/pytorch/issues/38349#issuecomment-627854342 for example

> Return value: For example, torch.nonzero returns a tensor when as_tuple=False (default) whereas the Data API requires to always return as tuple.

See #46187

> torch.cross looks for the first dimension found with the size 3, whereas the Data API uses -1 as default axis.

Cross is horrible, news at #50273

> floordiv operator needs be changed to take floor

This one is in progress #43874

**Comment by rgommers** [comment]
Opened the separate issue for DLPack: gh-55090

**Comment by asi1024** [comment]
@leofang @ezyang @rgommers Thank you for your response!
We updated [spreadsheet](https://docs.google.com/spreadsheets/d/13amsIeHhe38apCeZrOmsbsHT-aysmYur9-cugw5jVg0/edit?usp=sharing) accordingly and added links to each discussion issue.

Still that there are 3 remaining things (with BC) that haven't been discussed in issues

* `torch.Tensor.size()` should return the number of elements, not the shape. It also needs to be a property instead of a method.
* `torch.{max/min}` must always return `Tensor` instead of `torch.return_types.{max/min}`, even when `dim`/`keepdims` is specified.
* Python data API specifies that creation functions must return an array of floating data type, but the returned dtype of
  * `torch.arange` is `int64`.
  * `torch.full` is inferred from `fill_value` argument.

The former two issues are also better to have in terms of the NumPy compatibility, so we can discuss them in #38349. The last one is a Data API specific issue.

We are interested in contributing fixes needed to conforming PyTorch APIs to NumPy / Data API. Maybe we can start by implementing ones that are not compatibility breaking and do not require discussions, e.g., adding missing operators like `torch.Tensor.__pos__(x)` or supporting multiple arrays in `torch.result_type(*arrays)`?

**Comment by mruberry** [comment]
Hey @asi1024, thank you for helping with this analysis and wanting to help make PyTorch more compatible!

Implementing missing operators would be great and we will generally accept PRs for them. I suggest filing an issue for each missing operator first so we can briefly discuss the implementation plan and whether we think it's worthwhile to pursue.

If you would open an issue for the torch.{max/min} inconsistency that would be helpful, too. All these issues can cite this issue so it serves as "roll-up" for tracking purposes.

As for arange and full, PyTorch is consistent with NumPy's behavior of inferring the dtype. I'm surprised the data api would prefer the dtype not be inferred from the fill value. @rgommers, is that an oversight in the API, or was there some reason for this change?

Finally, it looks like the data API's discussion of tensor attributes is far from finalized, and changing the behavior of a function like size() would be challenging because of how heavily it's used. We may need to revisit that in the future, possibly by providing a wrapper for our tensors (or maybe the data API will provide a wrapper function like it provides the ability to query for a compatible namespace to use). An alternative would be for the API to add functions like *.size() and encourage calling them by using the compatible namespace.

Do those next steps make sense?

**Comment by asmeurer** [comment]
The arange question has been discussed at https://github.com/data-apis/array-api/issues/107.

**Comment by asi1024** [comment]
> Implementing missing operators would be great and we will generally accept PRs for them. I suggest filing an issue for each missing operator first so we can briefly discuss the implementation plan and whether we think it's worthwhile to pursue.

Great! I will start filing an issue first.

> If you would open an issue for the torch.{max/min} inconsistency that would be helpful, too.

Sure, will do.

> As for arange and full, PyTorch is consistent with NumPy's behavior of inferring the dtype.

I will start implementing non compatibility-breaking ones with specifications aligned between NumPy and Data API, and postpone ones with ongoing discussions in Data API (like `arange` and `size()`).

**Comment by mruberry** [comment]
> The arange question has been discussed at [data-apis/array-api#107](https://github.com/data-apis/array-api/issues/107).

I'm also confused by the discussion of arange, however (see https://github.com/data-apis/array-api/issues/107#issuecomment-758323129). How can floating point values be produced if start, end, and step are all integers?

**Comment by mruberry** [comment]
> @mruberry
> 
> > Implementing missing operators would be great and we will generally accept PRs for them. I suggest filing an issue for each missing operator first so we can briefly discuss the implementation plan and whether we think it's worthwhile to pursue.
> 
> Great! I will start filing an issue first.
> 
> > If you would open an issue for the torch.{max/min} inconsistency that would be helpful, too.
> 
> Sure, will do.
> 
> > As for arange and full, PyTorch is consistent with NumPy's behavior of inferring the dtype.
> 
> I will start implementing non compatibility-breaking ones with specifications aligned between NumPy and Data API, and postpone ones with ongoing discussions in Data API (like `arange` and `size()`).

Sounds great! Looking forward to seeing some issues!

**Comment by pmeier** [comment]
In an effort to "autogenerate" the analysis of @asi1024, I'm running a nightly PyTorch build against the provided [test suite of the array API specification](https://github.com/data-apis/array-api-tests). 

I'm tracking my progress on this branch: https://github.com/pmeier/array-api-tests/tree/pytorch. For a high level overview, see https://github.com/pmeier/array-api-tests/blob/pytorch/PYTORCH_README.md

**Comment by pmeier** [comment]
> How does that sound, @asi1024, @emcastillo, @pmeier?

I'll get on it.

> Note that we do have eigvalsh, inv, and pinv.

Yeah, but for convergence they need to be aliased to the `torch` namespace. This will be especially confusing for operators like [`torch.norm`](https://pytorch.org/docs/stable/generated/torch.norm.html?highlight=norm#torch.norm) which was just recently deprecated in favor of [`torch.linalg.norm`](https://pytorch.org/docs/stable/linalg.html?highlight=norm#torch.linalg.norm).

**Comment by rgommers** [comment]
> Yeah, but for convergence they need to be aliased to the `torch` namespace.

This is not correct, they need to be in the `linalg` namespace (and that's where these functions are also in numpy). Let me look at merging the PR that fixes this: https://github.com/data-apis/array-api/pull/182.

**Comment by pmeier** [comment]
> We should also create a new rollup issue with boxes for each of these (linked) subissues plus the DLpack issue @rgommers links above. Once we have the new rollup issue linking the subissues I think we can close this issue as a "Resolved RFC" and track progress and incompatibilities on the new tracker. We can call it the "Python Array API Compatibility Tracker", which is a long name, but oh well.

#58743

**Comment by mruberry** [comment]
Everyone OK with closing this as "RESOLVED: Support the Python Array API" and using the issue @pmeier filed to track progress implementing that support?

**Comment by mruberry** [comment]
Everyone OK with closing this as "RESOLVED: Support the Python Array API" and using the issue @pmeier filed to track progress implementing that support?

**Comment by rgommers** [comment]
> Everyone OK with closing this as "RESOLVED: Support the Python Array API" and using the issue @pmeier filed to track progress implementing that support?

Sounds good to me. I think it would be useful then to move over @asi1024's first comment and link to spreadsheet there - it has a lot of useful detail.

**Comment by asi1024** [comment]
@mruberry Sounds good to me!

**Comment by emcastillo** [comment]
Let's go ahead!! thank you everyone :)

**Comment by rgommers** [comment]
Seems like everyone is indeed happy, so closing.
